File: data-structures/heap/cpp/Heap.cpp
#include "Heap.h"
#include <iostream>

Heap::Heap() {}

Heap::~Heap() {}

void Heap::insert(int value) {
    heap.push_back(value);
    heapifyUp();
}

int Heap::getMax() const {
    if (!isEmpty()) {
        return heap[0];
    }
    return -1; // Heap is empty
}

int Heap::extractMax() {
    if (!isEmpty()) {
        int max = heap[0];
        heap[0] = heap.back();
        heap.pop_back();
        heapifyDown();
        return max;
    }
    return -1; // Heap is empty
}

bool Heap::isEmpty() const {
    return heap.empty();
}

void Heap::display() const {
    for (int value : heap) {
        std::cout << value << " ";
    }
    std::cout << std::endl;
}

void Heap::heapifyUp() {
    int index = heap.size() - 1;
    while (index > 0 && heap[index] > heap[(index - 1) / 2]) {
        std::swap(heap[index], heap[(index - 1) / 2]);
        index = (index - 1) / 2;
    }
}

void Heap::heapifyDown() {
    int index = 0;
    while (2 * index + 1 < heap.size()) {
        int leftChild = 2 * index + 1;
        int rightChild = 2 * index + 2;
        int largerChild = leftChild;

        if (rightChild < heap.size() && heap[rightChild] > heap[leftChild]) {
            largerChild = rightChild;
        }

        if (heap[index] < heap[largerChild]) {
            std::swap(heap[index], heap[largerChild]);
            index = largerChild;
        }
        else {
            break;
        }
    }
}

File: data-structures/heap/java/Heap.java
import java.util.ArrayList;
import java.util.List;

public class Heap {
    private List<Integer> heap;

    public Heap() {
        heap = new ArrayList<>();
    }

    private int parent(int i) {
        return (i - 1) / 2;
    }

    private int leftChild(int i) {
        return 2 * i + 1;
    }

    private int rightChild(int i) {
        return 2 * i + 2;
    }

    public void insert(int key) {
        heap.add(key);
        int i = heap.size() - 1;
        while (i != 0 && heap.get(parent(i)) < heap.get(i)) {
            swap(i, parent(i));
            i = parent(i);
        }
    }

    public int extractMax() {
        if (heap.size() == 0) {
            throw new IllegalStateException("Heap is empty");
        }
        if (heap.size() == 1) {
            return heap.remove(0);
        }

        int root = heap.get(0);
        heap.set(0, heap.remove(heap.size() - 1));
        maxHeapify(0);
        return root;
    }

    private void maxHeapify(int i) {
        int largest = i;
        int left = leftChild(i);
        int right = rightChild(i);

        if (left < heap.size() && heap.get(left) > heap.get(largest)) {
            largest = left;
        }
        if (right < heap.size() && heap.get(right) > heap.get(largest)) {
            largest = right;
        }

        if (largest != i) {
            swap(i, largest);
            maxHeapify(largest);
        }
    }

    private void swap(int i, int j) {
        int temp = heap.get(i);
        heap.set(i, heap.get(j));
        heap.set(j, temp);
    }

    public void display() {
        for (int i : heap) {
            System.out.print(i + " ");
        }
        System.out.println();
    }
}
